File: ui/src/main/java/org/wcl/mfn/ui/controller/tools/ContractCalculatorUIController.java
package org.wcl.mfn.ui.controller.tools;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.PropertySource;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.*;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.reactive.function.client.WebClient;
import org.wcl.mfn.api.controller.tools.ContractCalculatorAPIController;
import org.wcl.mfn.entities.contract.calculator.*;
import org.wcl.mfn.ui.config.common.PageTitleConfig;
import org.wcl.mfn.ui.config.mvc.ModelConfig;
import org.wcl.mfn.ui.config.mvc.ViewConfig;
import org.wcl.mfn.ui.config.tools.ContractCalculatorConfig;
import org.wcl.mfn.ui.config.tools.ContractCalculatorMessagesConfig;
import org.wcl.mfn.ui.controller.forms.contractcalculator.*;
import reactor.core.publisher.Mono;

import java.util.ArrayList;
import java.util.List;


@Controller
@PropertySource("classpath:/url/url.properties")
public class ContractCalculatorUIController {
    @Autowired
    private PageTitleConfig pageTitleConfig;

    @Autowired
    private ViewConfig viewConfig;

    @Autowired
    private ModelConfig modelConfig;

    @Autowired
    private ContractCalculatorConfig contractCalculatorConfig;

    @Autowired
    private ContractCalculatorMessagesConfig contractCalculatorMessagesConfig;

    @Autowired
    private ContractCalculatorAPIController apiController;

    private RequestedRemunerationForm requestedRemuneration = new RequestedRemunerationForm(0,0, new SalaryEscalatorForm(5));

    @GetMapping("/contract-calculator")
    public String contractCalculatorPage(Model model) {
        setModelAttributes(model, generateBlankContracts());

        return "pages/tools/contract-calculator";
    }

    @PostMapping("/contract-calculator/submit")
    public String updateSuggestedContracts(@ModelAttribute("requested_remuneration") RequestedRemunerationForm contractDemands, Model model) {
        requestedRemuneration = contractDemands;

        setModelAttributes(model, getSuggestedContracts());

        return "pages/tools/contract-calculator";
    }

    @GetMapping("/contract-calculator/reset")
    public String resetCalculatorPage(Model model) {
        setModelAttributes(model, generateBlankContracts());

        return "redirect:pages/tools/contract-calculator";
    }

    private void setModelAttributes(Model model, List<SuggestedContract> suggestedContracts) {
        model.addAttribute("page_title", pageTitleConfig.contractCalculatorPageTitle());
        model.addAttribute("requested_remuneration", requestedRemuneration);
        model.addAttribute("suggested_contracts", suggestedContracts);
    }

    private List<SuggestedContract> getSuggestedContracts() {
        final var headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.setAccept(List.of(MediaType.APPLICATION_JSON));

//        final var entity = new HttpEntity(new RequestedRemuneration(
//                requestedRemuneration.salary(),
//                requestedRemuneration.bonus()), headers);
        final var client = WebClient.builder()
                .baseUrl("http://localhost:8080/api/contract-calculator")
                .defaultHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
                .build();

        final Mono<List<SuggestedContract>> response = client.get()
                .accept(MediaType.APPLICATION_JSON)
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<>() {
                });

        return new ArrayList<>(response.block());
    }

    private List<SuggestedContract> generateBlankContracts()  {
        return List.of(new SuggestedContract(
                        2,0,0,0,
                        List.of(new SuggestedYearlyContract(1,0,0),
                                new SuggestedYearlyContract(2,0,0))
                ),
                new SuggestedContract(
                        3,0,0,0,
                        List.of(new SuggestedYearlyContract(1,0,0),
                                new SuggestedYearlyContract(2,0,0),
                                new SuggestedYearlyContract(3,0,0))
                ),
                new SuggestedContract(
                        4,0,0,0,
                        List.of(new SuggestedYearlyContract(1,0,0),
                                new SuggestedYearlyContract(2,0,0),
                                new SuggestedYearlyContract(3,0,0),
                                new SuggestedYearlyContract(4,0,0))
                ),
                new SuggestedContract(
                        5,0,0,0,
                        List.of(new SuggestedYearlyContract(1,0,0),
                                new SuggestedYearlyContract(2,0,0),
                                new SuggestedYearlyContract(3,0,0),
                                new SuggestedYearlyContract(4,0,0),
                                new SuggestedYearlyContract(5,0,0))
                ),
                new SuggestedContract(
                        6,0,0,0,
                        List.of(new SuggestedYearlyContract(1,0,0),
                                new SuggestedYearlyContract(2,0,0),
                                new SuggestedYearlyContract(3,0,0),
                                new SuggestedYearlyContract(4,0,0),
                                new SuggestedYearlyContract(5,0,0),
                                new SuggestedYearlyContract(6,0,0))
                )
        );
    }
}

File: ui/src/main/kotlin/org/wcl/mfn/ui/controller/tools/ContractCalculatorUIController.kt
package org.wcl.mfn.ui.controller.tools

import jakarta.validation.Valid
import jakarta.validation.constraints.Size
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.PropertySource
import org.springframework.core.ParameterizedTypeReference
import org.springframework.http.*
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.ModelAttribute
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.client.RestTemplate
import org.wcl.mfn.api.controller.tools.ContractCalculatorAPIController
import org.wcl.mfn.config.ui.common.PageTitleConfig
import org.wcl.mfn.config.ui.mvc.*
import org.wcl.mfn.config.ui.tools.*
import org.wcl.mfn.entities.contract.calculator.RequestedRemuneration
import org.wcl.mfn.entities.contract.calculator.SuggestedContract
import org.wcl.mfn.entities.contract.calculator.SuggestedYearlyContract
import org.wcl.mfn.ui.controller.forms.contractcalculator.RequestedRemunerationForm
import org.wcl.mfn.ui.controller.forms.contractcalculator.SalaryEscalatorForm
import java.net.URI


@Controller
@PropertySource("classpath:/url/url.properties")
class ContractCalculatorUIController {
    @Autowired
    private val pageTitleConfig: PageTitleConfig? = null

    @Autowired
    private val viewConfig: ViewConfig? = null

    @Autowired
    private val modelConfig: ModelConfig? = null

    @Autowired
    private val contractCalculatorConfig: ContractCalculatorConfig? = null

    @Autowired
    private val contractCalculatorMessagesConfig: ContractCalculatorMessagesConfig? = null

    @Autowired
    private val apiController: ContractCalculatorAPIController? = null

    private var requestedRemuneration = RequestedRemunerationForm(0,0, SalaryEscalatorForm(5))

    @GetMapping("/contract-calculator")
    fun contractCalculatorPage(model: Model): String {
        model.addAttribute("page_title",pageTitleConfig!!.contractCalculatorPageTitle())
        model.addAttribute("requested_remuneration", requestedRemuneration)
        model.addAttribute("suggested_contracts", generateBlankContracts())

        return "pages/tools/contract-calculator"
    }

    @PostMapping("/contract-calculator/submit")
    fun updateSuggestedContracts(@ModelAttribute("requested_remuneration") contractDemands:RequestedRemunerationForm, model:Model): String? {
        requestedRemuneration = contractDemands;

        model.addAttribute("page_title",pageTitleConfig!!.contractCalculatorPageTitle())
        model.addAttribute("requested_remuneration", requestedRemuneration)
        model.addAttribute("suggested_contracts", getSuggestedContracts())

        return "pages/tools/contract-calculator"
    }

    private fun getSuggestedContracts() : List<SuggestedContract> {
        val headers = HttpHeaders()
        headers.contentType = MediaType.APPLICATION_JSON
        headers.accept = listOf(MediaType.APPLICATION_JSON)

        val entity: HttpEntity<RequestedRemuneration> = HttpEntity(RequestedRemuneration(
            requestedRemuneration.salary,
            requestedRemuneration.bonus), headers)

        val restTemplate = RestTemplate()

        val response: ResponseEntity<List<SuggestedContract>> = restTemplate.exchange(
            URI.create("http://localhost:8080/api/contract-calculator"),
            HttpMethod.POST,
            entity,
            typeReference<List<SuggestedContract>>())
        return response.body
    }

    private fun generateBlankContracts(): List<SuggestedContract> {
        return listOf(SuggestedContract(
            2,0,0,0,
            listOf(SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0))
            ),
            SuggestedContract(
                3,0,0,0,
                listOf(SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0))
            ),
            SuggestedContract(
                4,0,0,0,
                listOf(SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0))
            ),
            SuggestedContract(
                5,0,0,0,
                listOf(SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0))
            ),
            SuggestedContract(
                6,0,0,0,
                listOf(SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0),
                    SuggestedYearlyContract(0,0,0))
            )
        )
    }
//
//    internal data class ContractParameters(val salary: Int, val bonus: Int, val escalator: Int) {
//        constructor(salary: Int, bonus: Int) : this(salary, bonus, DEFAULT_ESCALATOR)
//
//        companion object {
//            private const val DEFAULT_ESCALATOR = 5
//        }
//    }

//    internal data class Escalator(val label: String?, val defaultValue: Int, val max: Int, val step: Int)

    private inline fun <reified T> typeReference() = object : ParameterizedTypeReference<T>() {}
}


